packages/atclient/src/connection.c

```diff
@@ -204,29 +202,77 @@ exit: {
 }
 }
 
-int atclient_connection_send(atclient_connection *ctx, const unsigned char *src, const size_t srclen,
-                             unsigned char *recv, const size_t recvsize, size_t *recvlen) {
+int atclient_connection_send(atclient_connection *ctx, const unsigned char *src_r, const size_t srclen_r,
+                             unsigned char *recv, const size_t recvsize_r, size_t *recvlen) {
   int ret = 1;
 
+  // Clone readonly inputs so it is editable by the hooks
+  size_t srclen = srclen_r;
+  size_t recvsize = recvsize_r;
+
+  bool try_hooks = ctx->hooks != NULL && !ctx->hooks->_is_nested_call;
+  bool allocate_src = try_hooks && ctx->hooks->readonly_src == false;
+
+  unsigned char *src;
+
+  if (allocate_src) {
+    src = malloc(sizeof(unsigned char) * srclen);
+    if (src == NULL) {
+      atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Failed to allocate memory for src\n");
+      allocate_src = false; // don't try to free since the memory failed to be allocated
+      goto exit;
+    }
+    memcpy(src, src_r, srclen);
+  } else {
+    src = (unsigned char *)src_r;
+  }
+
```

Comment:
src is conditionally copied & allocated as an editable buffer, but only if readonly_src = false. This allows us to maintain performance by default, but allows the end developer to take control if they really want to.